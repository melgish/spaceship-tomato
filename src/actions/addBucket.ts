'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddBucket, addBucketSchema } from '@/schemas/addBucket.schema'

/**
 * Add server action to a new bucket to the database.
 * @see ../schemas/addBucket.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addBucket(payload: AddBucket) {
  try {
    const data = await addBucketSchema.parseAsync(payload)
    const created = await prisma.bucket.create({
      data: { ...data, plantId: data.plantId || null },
    })
    revalidatePath('/buckets')
    return { ok: true, message: 'Bucket created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create bucket' } as const
  }
}
