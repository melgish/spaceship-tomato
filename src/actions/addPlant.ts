'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddPlant, addPlantSchema } from '@/schemas/addPlant.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addPlant.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addPlant(payload: AddPlant) {
  try {
    const data = await addPlantSchema.parseAsync(payload)
    const created = await prisma.plant.create({ data })
    revalidatePath('/plants')
    return { ok: true, message: 'Plant created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create plant' } as const
  }
}
