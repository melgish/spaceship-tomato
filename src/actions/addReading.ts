'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddReading, addReadingSchema } from '@/schemas/addReading.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addReading.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addReading(payload: AddReading) {
  try {
    const data = await addReadingSchema.parseAsync(payload)
    const created = await prisma.reading.create({ data })
    revalidatePath('/readings')
    return { ok: true, message: 'Reading created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create reading' } as const
  }
}
