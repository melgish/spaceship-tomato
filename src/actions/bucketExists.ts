'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that a bucket exists in the database.
 * @param id Id of bucket to search for.
 * @returns true if bucket exists, otherwise false.
 */
export async function bucketExists(id: number = 0) {
  return id > 0 && (await prisma.bucket.count({ where: { id } })) > 0
}
