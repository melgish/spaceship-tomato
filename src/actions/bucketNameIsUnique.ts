'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that the name of a bucket is unique.
 * @param name Name of the bucket to test.
 * @param id Optional id of bucket to exclude when editing an existing bucket.
 * @returns true if the name exists on another bucket, otherwise false.
 */
export async function bucketNameIsUnique(name: string, id: number = 0) {
  const where = { name, id: (id > 0 && { not: id }) || undefined }
  return (await prisma.bucket.count({ where })) === 0
}
