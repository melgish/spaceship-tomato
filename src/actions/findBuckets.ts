'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all of buckets from the database.
 * @returns An array of buckets. The returned data will include plant info if
 * the bucket contains a plant.
 */
export async function findBuckets() {
  return await prisma.bucket.findMany({
    include: { plant: true },
    orderBy: [{ name: 'asc' }],
  })
}

export type Buckets = Awaited<ReturnType<typeof findBuckets>>
export type Bucket = Buckets[number]
