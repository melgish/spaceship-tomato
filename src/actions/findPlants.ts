'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all of plants from the database.
 * @returns An array of plants.
 */
export async function findPlants() {
  return await prisma.plant.findMany({
    orderBy: [{ name: 'asc' }],
  })
}

export type Plants = Awaited<ReturnType<typeof findPlants>>
export type Plant = Plants[number]
