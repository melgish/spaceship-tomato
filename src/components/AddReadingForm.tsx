'use client'
import type { Buckets } from '@/actions/findBuckets'

import { useId } from 'react'
import { useForm } from 'react-hook-form'
import { zodResolver } from '@hookform/resolvers/zod'
import ErrorMessage from '@/components/ErrorMessage'
import ResetButton from '@/components/ResetButton'
import SubmitButton from '@/components/SubmitButton'

import { addReading } from '@/actions/addReading'
import { type AddReading, addReadingSchema } from '@/schemas/addReading.schema'
import type { IdAndName } from '@/lib/idAndName'
import { notify } from '@/lib/notify'

const defaultValues: AddReading = { bucketId: -1, ec: 0, ph: 0, tds: 0 }
const resolver = zodResolver(addReadingSchema)

type Props = Readonly<{
  buckets: Buckets
  bucketId?: number
}>

export default function AddReadingForm({ buckets, bucketId = 0 }: Props) {
  const id = useId()

  const clear = () => {
    reset({ ...defaultValues, bucketId })
  }

  const save = async (data: AddReading) => {
    if (notify(await addReading(data))) {
      clear()
    }
  }

  const { handleSubmit, formState, register, reset } = useForm<AddReading>({
    defaultValues: { ...defaultValues, bucketId },
    resolver,
    shouldFocusError: true,
  })

  return (
    <form className="flex gap-x-3" noValidate onSubmit={handleSubmit(save)}>
      <div className="required field">
        <label htmlFor={`${id}-bucketId`}>Bucket</label>
        <select
          {...register('bucketId')}
          className="py-1.5"
          style={{ width: '20ch' }}
          id={`${id}-bucketId`}
          disabled={bucketId > 0}
        >
          <option value={0}>None</option>
          {buckets.map((p: IdAndName) => (
            <option key={p.id} value={p.id}>
              {p.name}
            </option>
          ))}
        </select>
        <ErrorMessage error={formState.errors?.bucketId} />
      </div>

      <div className="required field">
        <label htmlFor={`${id}-ph`}>PH</label>
        <input
          {...register('ph')}
          id={`${id}-ph`}
          max={14}
          min={0}
          placeholder="Ex. 7.0"
          required
          size={10}
          step=".1"
          type="number"
        />
        <ErrorMessage error={formState.errors?.ph} />
      </div>

      <div className="required field">
        <label htmlFor={`${id}-ec`}>EC (&micro;S/cm)</label>
        <input
          {...register('ec')}
          id={`${id}-ec`}
          max={14}
          min={0}
          placeholder="Ex. 7.0"
          required
          size={10}
          step=".1"
          type="number"
        />
        <ErrorMessage error={formState.errors?.ec} />
      </div>

      <div className="required field">
        <label htmlFor={`${id}-tds`}>TDS (ppm)</label>
        <input
          {...register('tds')}
          id={`${id}-tds`}
          max={14}
          min={0}
          placeholder="Ex. 7.0"
          required
          size={10}
          step=".1"
          type="number"
        />
        <ErrorMessage error={formState.errors?.tds} />
      </div>

      <div className="flex items-center gap-x-2">
        <SubmitButton />
        <ResetButton onClick={clear} />
      </div>
    </form>
  )
}
