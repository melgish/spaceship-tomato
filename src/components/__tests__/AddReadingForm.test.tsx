import { Mock, beforeEach, describe, expect, it, vi } from 'vitest'
import { act, fireEvent, render, screen } from '@testing-library/react'
import AddReadingForm from '../AddReadingForm'
import { addReading } from '@/actions/addReading'
import { bucketExists } from '@/actions/bucketExists'

const buckets = [
  { id: 1, name: 'Bucket 1', plantId: 1, plant: { name: 'Plant 1' } },
  { id: 2, name: 'Bucket 2', plantId: 0, plant: null },
] as any

vi.mock('@/actions/bucketExists')
vi.mock('@/actions/addReading')
vi.mock('react-hot-toast')

beforeEach(() => {
  // Mock deep validation
  ;(bucketExists as Mock).mockResolvedValue(true)
  ;(addReading as Mock).mockResolvedValue({ ok: true, message: 'Good' })
})

it('should render', () => {
  const { container } = render(<AddReadingForm buckets={buckets} />)

  expect(container).toMatchSnapshot()
})

describe('when invalid and submit is clicked', () => {
  it('should report validation errors', async () => {
    const { container } = render(<AddReadingForm buckets={buckets} />)

    const submit = screen.getByRole('button', { name: /submit/i })
    await act(() => fireEvent.click(submit))

    expect(container).toMatchSnapshot()
    expect(addReading).not.toBeCalled()
  })
})

describe('when valid and submit is clicked', () => {
  it('should submit data', async () => {
    const { container } = render(<AddReadingForm buckets={buckets} />)

    const bucket = screen.getByLabelText(/bucket/i)
    const setBucket = { target: { value: 1 } }
    await act(() => fireEvent.change(bucket, setBucket))

    const ph = screen.getByLabelText(/ph/i)
    const setPH = { target: { value: '6.2', valueAsNumber: 6.2 } }
    await act(() => fireEvent.change(ph, setPH))

    const ec = screen.getByLabelText(/ec/i)
    const setEC = { target: { value: '1234', valueAsNumber: 1234 } }
    await act(() => fireEvent.change(ec, setEC))

    const tds = screen.getByLabelText(/tds/i)
    const setTDS = { target: { value: '875', valueAsNumber: 875 } }
    await act(() => fireEvent.change(tds, setTDS))

    const submit = screen.getByRole('button', { name: /submit/i })
    await act(() => fireEvent.click(submit))

    expect(container).toMatchSnapshot()
    expect(addReading).toBeCalledWith({
      bucketId: 1,
      ph: 6.2,
      tds: 875,
      ec: 1234,
    })
  })
})
