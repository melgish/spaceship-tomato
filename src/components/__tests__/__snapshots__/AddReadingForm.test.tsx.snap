// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render 1`] = `
<div>
  <form
    class="flex gap-x-3"
    novalidate=""
  >
    <div
      class="required field"
    >
      <label
        for=":r0:-bucketId"
      >
        Bucket
      </label>
      <select
        class="py-1.5"
        id=":r0:-bucketId"
        name="bucketId"
        style="width: 20ch;"
      >
        <option
          value="0"
        >
          None
        </option>
        <option
          value="1"
        >
          Bucket 1
        </option>
        <option
          value="2"
        >
          Bucket 2
        </option>
      </select>
    </div>
    <div
      class="required field"
    >
      <label
        for=":r0:-ph"
      >
        PH
      </label>
      <input
        id=":r0:-ph"
        max="14"
        min="0"
        name="ph"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
    </div>
    <div
      class="required field"
    >
      <label
        for=":r0:-ec"
      >
        EC (µS/cm)
      </label>
      <input
        id=":r0:-ec"
        max="14"
        min="0"
        name="ec"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
    </div>
    <div
      class="required field"
    >
      <label
        for=":r0:-tds"
      >
        TDS (ppm)
      </label>
      <input
        id=":r0:-tds"
        max="14"
        min="0"
        name="tds"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
    </div>
    <div
      class="flex items-center gap-x-2"
    >
      <button
        class="emerald"
        type="submit"
      >
        Submit
      </button>
      <button
        class="slate"
        type="button"
      >
        Clear
      </button>
    </div>
  </form>
</div>
`;

exports[`when invalid and submit is clicked > should report validation errors 1`] = `
<div>
  <form
    class="flex gap-x-3"
    novalidate=""
  >
    <div
      class="required field"
    >
      <label
        for=":r1:-bucketId"
      >
        Bucket
      </label>
      <select
        class="py-1.5"
        id=":r1:-bucketId"
        name="bucketId"
        style="width: 20ch;"
      >
        <option
          value="0"
        >
          None
        </option>
        <option
          value="1"
        >
          Bucket 1
        </option>
        <option
          value="2"
        >
          Bucket 2
        </option>
      </select>
      <div
        class="error text-rose-800"
      >
        A bucket is required
      </div>
    </div>
    <div
      class="required field"
    >
      <label
        for=":r1:-ph"
      >
        PH
      </label>
      <input
        id=":r1:-ph"
        max="14"
        min="0"
        name="ph"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
    </div>
    <div
      class="required field"
    >
      <label
        for=":r1:-ec"
      >
        EC (µS/cm)
      </label>
      <input
        id=":r1:-ec"
        max="14"
        min="0"
        name="ec"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
      <div
        class="error text-rose-800"
      >
        EC &gt;= 1
      </div>
    </div>
    <div
      class="required field"
    >
      <label
        for=":r1:-tds"
      >
        TDS (ppm)
      </label>
      <input
        id=":r1:-tds"
        max="14"
        min="0"
        name="tds"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
      <div
        class="error text-rose-800"
      >
        TDS &gt;= 1
      </div>
    </div>
    <div
      class="flex items-center gap-x-2"
    >
      <button
        class="emerald"
        type="submit"
      >
        Submit
      </button>
      <button
        class="slate"
        type="button"
      >
        Clear
      </button>
    </div>
  </form>
</div>
`;

exports[`when valid and submit is clicked > should submit data 1`] = `
<div>
  <form
    class="flex gap-x-3"
    novalidate=""
  >
    <div
      class="required field"
    >
      <label
        for=":r2:-bucketId"
      >
        Bucket
      </label>
      <select
        class="py-1.5"
        id=":r2:-bucketId"
        name="bucketId"
        style="width: 20ch;"
      >
        <option
          value="0"
        >
          None
        </option>
        <option
          value="1"
        >
          Bucket 1
        </option>
        <option
          value="2"
        >
          Bucket 2
        </option>
      </select>
    </div>
    <div
      class="required field"
    >
      <label
        for=":r2:-ph"
      >
        PH
      </label>
      <input
        id=":r2:-ph"
        max="14"
        min="0"
        name="ph"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
    </div>
    <div
      class="required field"
    >
      <label
        for=":r2:-ec"
      >
        EC (µS/cm)
      </label>
      <input
        id=":r2:-ec"
        max="14"
        min="0"
        name="ec"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
    </div>
    <div
      class="required field"
    >
      <label
        for=":r2:-tds"
      >
        TDS (ppm)
      </label>
      <input
        id=":r2:-tds"
        max="14"
        min="0"
        name="tds"
        placeholder="Ex. 7.0"
        required=""
        size="10"
        step=".1"
        type="number"
      />
    </div>
    <div
      class="flex items-center gap-x-2"
    >
      <button
        class="emerald"
        type="submit"
      >
        Submit
      </button>
      <button
        class="slate"
        type="button"
      >
        Clear
      </button>
    </div>
  </form>
</div>
`;
